### DISCUSSION TITLE
Referenced call of logger.info() or logger.error()

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#5019] [Q&A] [answered]
Best description for my request is a code snippet:
```js
// Depending on a condition I want to call either logger.info() or logger.error()
// ... so I store a reference to that function in variable `func` as follows:
var func;
if(1) {
	func = logger.info;
} else {
	func = logger.error;
}
logger.debug("func: " + func);
```
Prints:
```
DEBUG  (MyChannel:deploy:?): func: function info() {/*
void info(java.lang.Object,java.lang.Throwable)
void info(java.lang.Object)
*/}
```
Looks great!

Now call it:
```js
func("foo");
```
Fails with error: `Java method "info" was invoked with [object global] as "this" value that can not be converted to Java type org.apache.log4j.Category.`

# How to solve that?

Full error:
```
ERROR  (com.mirth.connect.server.util.javascript.JavaScriptUtil:522): Error executing 1699860e-d21a-4d8b-b6e8-b6853b32b767_Deploy script from channel: 1699860e-d21a-4d8b-b6e8-b6853b32b767
com.mirth.connect.server.MirthJavascriptTransformerException: CHANNEL:
	MyChannelSOURCE CODE:
	2460: 
	func = logger.info;2461: } else {2462: 
	func = logger.error;2463: }2464: logger.debug("func: " + func);2465: func("foo");2466:  2467: }LINE NUMBER:
	2465DETAILS:
	Java method "info" was invoked with [object global] as "this" value that can not be converted to Java type org.apache.log4j.Category.
	at 1699860e-d21a-4d8b-b6e8-b6853b32b767_Deploy:2465 (doScript)
	at 1699860e-d21a-4d8b-b6e8-b6853b32b767_Deploy:2468
	at com.mirth.connect.server.util.javascript.JavaScriptUtil.executeScript(JavaScriptUtil.java:547)
	at com.mirth.connect.server.util.javascript.JavaScriptUtil$2.doCall(JavaScriptUtil.java:379)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:113)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by tom08zehn** [comment]
P.S. The following answer is not accepted 😅
```js
if(1) {
	logger.info("foo");
} else {
	logger.error("foo");
}
```
The background of my request is more complex and the code snippet just illustrates the basic issue.

**Comment by jonbartels** [comment]
I'm not sure of the answer but this information might help.

The logger is added to the JS scope here `com.mirth.connect.server.util.javascript.JavaScriptScopeUtil#addLogger`. The `logger` it adds seems to come from a call to `org.apache.log4j.Logger#getLogger(java.lang.String)` . So when you define `func` you are adding a JS reference to a Java method (I think??). That then tries to call `org.apache.log4j.Category#info(java.lang.Object)` which doesn't have the right context for `this`; I don't really understand how `this` context works in Rhino though. I am stuck at this point.

Does this information help you?

What is your use case? Do you just want a universal toggle between logging at the info level and the error level?

**Reply by tom08zehn** [reply]
Yes, I understand the same. And maybe the problem is that `func("foo")` submits a JavaScript string to the referenced Java method and might be converted from JS string to a Rhino (= Java) object, what is not a valid argument type.

I will try `func(new java.lang.String("foo"))` and report the result.

**Reply by tom08zehn** [reply]
not working, same error

**Comment by odoodo** [comment]
Updated (and this time tested) proposal:
```js
var func;
if(1) {
	func = function(something){logger.info(something)};
} else {
	func = function(something){logger.error(something)};
}
// first parameter is the context 
func.call(this, 'I am some log message');
```

Further details about call() can be found [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call).

**Reply by tom08zehn** [reply]
not working, error: `Can't find method org.apache.log4j.Category.debug().`

**Reply by jonbartels** [reply]
@tom08zehn what if you pass args? `func.call("foo", new java.lang.String("odoodo is awesome!"));`

**Reply by tom08zehn** [reply]
error `Java method "debug" was invoked with foo as "this" value that can not be converted to Java type org.apache.log4j.Category.`

**Reply by odoodo** [reply]
Yes, you are right, my proposal wasn't complete. Please find updated code in my first post. Sorry for the confusion.

The reason for the error messages is that the Category object does not possess a static functions for debug() or info() etc.
That's why I had to wrap the calls into anonymous functions.

**Comment by tonygermano** [comment, marked as answer]
In pure javascript, when you invoke `logger.info("foo")` it is (roughly) the same as `logger.info.call(logger, "foo")`.

When you create the reference to the function at the top level, you are no longer calling it as a property of an object, so it is roughly the same as `logger.info.call(this, "foo")`, where `this` is the global object. Since info is really a Java method, it's going to complain if you try to invoke it on the wrong type.

Using Function.prototype.bind, you can create a new function that always uses the same object for the value of `this` when the method is invoked.

```js
var func;
if(1) {
	func = logger.info.bind(logger);
} else {
	func = logger.error.bind(logger);
}
func("foo")
```

**Reply by tom08zehn** [reply]
Great, thanks! Final result:
```js
var func;

if(1) {
	func = logger.info;
} else {
	func = logger.error;
}

func = func.bind(logger);
func("foo");

/* or as a one-liner:
func.bind(logger)("foo");
*/
```

**Reply by tonygermano** [reply]
Each time you call `bind` it returns a new function, so if you're going to be calling `func` multiple times, I'd make it a reference to the bound function to begin with.

There's no reason to call it as a one-liner, as `func.call(logger, "foo")` will accomplish the same thing without creating the new function.
